File: 113bN_EC_proper_event_summary.R
######
# Adam Griffin, 2020-04-27
#
# Summarising size and time, and extracting daily PoE of extreme events.
# 
# For aquaCAT, Project 07441.
# 
# Created ABG 2020-04-27
# Pipeline version ABG 2020-09-07
# Netcdf version ABG 2021-07-07
#
#
# OUTPUTS: eventdf_***.csv: dataframe of each event summarised. One event per row.
#
#####

if (interactive()) {commandArgs <- function(...){c("05","present")}}
#### SETUP ####----------------------
if (substr(osVersion,1,3) == "Win") {
  source("S:/CodeABG/setup_script_00.R")
}else if (substr(osVersion,1,3) %in% c("Cen","Fed")){
  source("/prj/aquacat/CodeABG/setup_script_00.R")
}else{
  source("~/AQUACAT/CodeABG/setup_script_00.R")
}

if (settings$EC2summ) {
  stop("EC summary exists. Stopping 113bN.")
}

print(paste("Running for threshold", thresh1, "at", ws1, "minimum spread."))

# matrix of threshold value (col) at a given cell (row)
threshMat <- readRDS(paste0(data_wd, subfold, "threshMat_RCM",
                            RCM, suffix, ".rds"))
#dim(threshMat) #19914 x 5

print("Loading in data")
load(paste0(data_wd, subfold, "eventLists_RCM", RCM, suffix, ".RDa"))

param_table <- readdf(paste0(data_wd, subfold, "paramtableG_POT2_RCM",
                            RCM, suffix, ".csv"))

#### EC SUMMARY ####

ec_events <- nc_open(paste0(data_wd,subfold, "eventEC_",
                            thresh1,"_", ws1,"_RCM", RCM, suffix, ".nc"))

eventNo <- ncvar_get(ec_events, "eventNo")
NE <- NN <- sum(eventNo > 0)

eventSummaryFrame <- data.frame(eventNumber=numeric(NN),
                                eventDay=numeric(NN),
                                eventLength = numeric(NN),
                                area = numeric(NN),
                                peakA = numeric(NN),
                                peakD = numeric(NN),
                                season = character(NN),
                                nclusters = numeric(NN),
                                peakyness = numeric(NN),
                                stringsAsFactors = FALSE)

# (D %/% 90) %% 4 for season 0,1,2,3: DJF, MAM, JJA, SON
season <- function(D) switch(((D %/% 90) %% 4) + 1, "DJF", "MAM", "JJA", "SON")

nclustevent <- function(v, radii=150) {
  # This function estimates the number of "seperate clusters" of event
  # going on during a widespread event. 
  ne0 <- sum(v, na.rm=T)
  event1 <- rbind(as.matrix(rn[v,]), matrix(c(
    700, 0, 700000, 1000000
  ), nrow = 1))
  d <- dist(event1[, 1:2])
  d2 <- dist(event1[1:ne0, 1:2])
  if ((max(d2) > radii) & (ne0 > 1)) {
    hh <- hclust(d, method = "single")
    ch <- max(cutree(hh, h = 0.8 * sort(hh$height, decreasing = T)[2])) - 1
    if(sort(hh$height,decreasing=T)[2] < 0.2*max(hh$height)){
      ch <- 1
    }
    ch <- min(ch, 4)
  } else{
    ch <- 1
  }
  ch
}

peakyfun <- function(v, thr) {
  mn <- min(v)
   if (all(is.na(v))| mn > thr) {
    return(0)
  }else{
    mn50 <- mean(c(mn, thr))
    return(sum(v < mn50) / sum(v < thr))
  }
}

rn <- as.data.frame(rn)
ST <- Sys.time()
ST0 <- Sys.time()

for(i in 1:NE){
  if((i < 10) | (i %% 400 == 0)){ # time recording
    print(i)
    I <- difftime(Sys.time(), ST, units="secs")
    I0 <- difftime(Sys.time(), ST0, units="secs")
    print(paste("Percent remaining", 100*round((NN-i)/NN ,2)))
    print(paste("Time remaining", round((NN-i)/i * I0,2)))
    print(paste("Since last readout:", round(I,2)))
    ST <- Sys.time()
    print(ST) 
  }
  eventSummaryFrame[i,] <- eventSummaryLine(ec_events, i, param_table$threshold)
}

print("saving EC Summary")
readr::write_csv(eventSummaryFrame,
                 paste0(data_wd,subfold, "eventSumm_ECX_",
                        thresh1,"_", ws1,"_RCM", RCM, suffix, ".csv"))

settings$EC2summ <- TRUE
write_yaml(settings, settingspath)

print("113bN done.")

File: 113bN_EC_proper_event_summary.py
# -*- coding: utf-8 -*-
"""
Created on Thu Sep  2 11:17:32 2021
Author: Adam Griffin, UKCEH
Project: AQUACAT
Script to summarise EC events.
"""
import os
import netCDF4 as nc
import pandas as pd
import numpy as np
import re
import sys
import yaml
import time
import gc

def season(x):
  return ["DJF","MAM","JJA","SON"][((x // 90) % 4)]

#rcm = "01"
#period="205012_208011"

rcm = sys.argv[1]

if sys.argv[2] == "present":
    period = "198012_201011"
else:
    period = "205012_208011"

print(f"Running RCM {rcm} for {period}.")

RCMS = ["01","04","05","06","07","08","09","10","11","12","13","15"]
periods = ["198012_201011","205012_208011"]

# CHANGE THIS TO THE TOP LEVEL OF THE FOLDER THE CSVs ARE IN
toplevel = r"/prj/aquacat/Data" #r'S:/Data'

# CHANGE THIS TO THE TOP LEVEL OF THE FOLDER THE NETCDFs ARE IN
outlevel = toplevel #r'S:/Data'

# CHANGE THIS TO WHERE THE hasData files are, they should exist in the toplevel folder.
rn = pd.read_csv(f"{toplevel}/hasData_primary.csv")
rnreg = pd.read_csv(f"{toplevel}/hasData_Regions.csv")

method="EC"

regional = False

if regional:
    subfold='/NW'
    fileinfix = 'NW_POT2_pc01'
    rn = rn[rnreg.REGION=="NW"]
    NH = 1437
else:
    subfold=''
    fileinfix = 'POT2_pc01'
    NH = 19914

if len(sys.argv) > 3:
    if sys.argv[3] == "FF":
        subfold = '_FF'

ncpath = (f"{outlevel}/RCM{rcm}_{period}{subfold}/event{method}"
          f"_POT2_pc01_RCM{rcm}_{period}.nc")
ncfile = nc.Dataset(ncpath, mode='r')

param_path = (f"{outlevel}/RCM{rcm}_{period}{subfold}/paramtableG_POT2_RCM{rcm}"
              f"_{period}.csv")

param_table = pd.read_csv(param_path)

thresh_path = f"{outlevel}/RCM{rcm}_{period}{subfold}/threshMat_RCM{rcm}_{period}.csv"

threshvec = pd.read_csv(thresh_path).iloc[:,1]
if regional:
    threshvec = threshvec[rnreg.REGION == "NW"]

summ_path = (f"{outlevel}/RCM{rcm}_{period}{subfold}/eventSumm_OBS"
             f"_POT2_pc01_RCM{rcm}_{period}.csv")
summtable = pd.read_csv(summ_path)

init_path = (f"{outlevel}/RCM{rcm}_{period}{subfold}/initialSummary"
             f"_RCM{rcm}_{period}.csv")
init_table = pd.read_csv(init_path)

summtable_out = pd.DataFrame(columns=["eventNumber", "eventDay", "eventLength",
                                      "area","peakA", "peakA_mid", "peakD", "season",
                                      "nclusters","peakyness"])

eventNo = list(ncfile.variables["eventNo"][:])
NE = np.sum([i > 0 for i in eventNo])

avec_all = ncfile.variables['ape'][:,:]
avec_mid = ncfile.variables['ape_mid'][:,:]

print("Setup complete")

start_time = time.time()
for i in range(NE):
    if (i < 10) or (i % 1000) == 0:
        print(i)
        print("--- %s seconds ---" % (time.time() - start_time))
        start_time = time.time()
    ni = eventNo[i]
    vvec = 0
    avec = min(avec_all[i,:])
    amid = min(avec_mid[i,:])
    dvec = 0
    D = init_table.iloc[ni-1,:] # Done in R afterwards
    #seas = init_table.iloc[ni-1, 3]) # Done in R afterwards
    summtable_out.loc[i] = [ni, D[0], D[1],
                            vvec, avec, amid, dvec,
                            D[3], 0, 0]

print("--- %s seconds ---" % (time.time() - start_time))
print("avec amid done")

del avec_all
del avec_mid
gc.collect()

vvec_all = ncfile.variables['flow'][:,:]
dvec_all = ncfile.variables['dpe'][:,:]

for i in range(NE):
    if (i < 10) or (i % 1000) == 0:
        print(i)
        print(">>> %s seconds >>>" % (time.time() - start_time))
        start_time = time.time()
    ni = eventNo[i]
    summtable_out.iloc[i,3] = sum(vvec_all[i,:] > threshvec)
    summtable_out.iloc[i,6] = min(dvec_all[i,:])



print(">>> %s seconds >>>" % (time.time() - start_time))
print("vvec dvec done")


ncfile.close()

yaml_path = f"{outlevel}/RCM{rcm}_{period}/settings.yaml"

summpath_out = (f"{outlevel}/RCM{rcm}_{period}{subfold}/eventSumm_"
                f"{method}_POT2_pc01_RCM{rcm}_{period}.csv")

summtable_out.to_csv(summpath_out, index=False) 

with open(yaml_path) as ym:
    list_doc = yaml.safe_load(ym)

list_doc['EC2summ'] = True
list_doc['ECpropsumm'] = "113bN.py"

with open(yaml_path, 'w') as ym:
    yaml.dump(list_doc, ym)

print(time.strftime("%Y-%m-%d %H:%M:%S"))
print("Files saved and YAML updated. End.")
